public/app/features/dashboard/components/ShareModal/SharePublicDashboard/SharePublicDashboard.tsx

```diff
@@ -80,10 +88,13 @@ export const SharePublicDashboard = (props: Props) => {
   const onSavePublicConfig = () => {
     reportInteraction('grafana_dashboards_public_create_clicked');
 
-    saveConfig({
+    const req = {
       dashboard: props.dashboard,
       payload: { ...publicDashboard!, isEnabled: enabledSwitch.isEnabled, annotationsEnabled },
-    });
+    };
+
+    // create or update based on whether we have existing uid
+    props.dashboard.meta.hasPublicDashboard ? updatePublicDashboard(req) : createPublicDashboard(req);
```

Comment:
I don't like to put this logic inside here. Maybe we can create a custom hook like useUpsertPublicDashboard() and expose an upsert method which inside decides if it should be an insert or an update call.

Reply:
That would create the same problem we had with public dashboard api where we have a catchall method that splits between the two code paths and adds complexity as we evolve. We want to handle this at the time the decision is being made. Added benefit is not needing to pass in additional dependencies (such as dashboard meta)